### PR TITLE
 Introduce private_key_jwt auth client handler in identity-oauth-addons

### PR DESCRIPTION
**Pull request by ayshsandu**
## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/PrivateKeyJWTClientAuthHandler.java [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+    }
+
+    @Override
+    public boolean canAuthenticate(OAuthTokenReqMessageContext tokReqMsgCtx) throws IdentityOAuth2Exception {
+
+        String clientAssertionType = getRequestParameter(tokReqMsgCtx, Constants.OAUTH_JWT_ASSERTION_TYPE);
+        String clientAssertion = getRequestParameter(tokReqMsgCtx, Constants.OAUTH_JWT_ASSERTION);
+
+        return isValidJWTClientAssertionRequest(clientAssertionType, clientAssertion);
+        //TODO check validate SAML_Bearer
```
Comment (sagara-gunathunga) [inline component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/PrivateKeyJWTClientAuthHandler.java]: Why we have this TODO here ?

File: component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/PrivateKeyJWTClientAuthHandler.java [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+    }
+
+    @Override
+    public boolean authenticateClient(OAuthTokenReqMessageContext tokReqMsgCtx)
+            throws IdentityOAuth2Exception {
+        try {
+            SignedJWT signedJWT = getSignedJWT(tokReqMsgCtx);
+            return jwtValidator.isValidToken(signedJWT);
+        } catch (IdentityOAuth2Exception e) {
+            return false;
```
Comment (sagara-gunathunga) [inline component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/PrivateKeyJWTClientAuthHandler.java]: We need to handle this IdentityOAuth2Exception, if it's intentional please explain  through a comment

File: component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/storage/JWTStorageManager.java [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+package org.wso2.carbon.identity.oauth2.token.handler.clientauth.jwt.storage;
+
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.identity.core.util.IdentityDatabaseUtil;
+import org.wso2.carbon.identity.oauth2.IdentityOAuth2Exception;
+import org.wso2.carbon.identity.oauth2.token.handler.clientauth.jwt.Constants;
+
+import java.sql.*;
```
Comment (sagara-gunathunga) [inline component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/storage/JWTStorageManager.java]: ?

Comment (ayshsandu) [reply component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/storage/JWTStorageManager.java]: fixed.

File: component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/PrivateKeyJWTClientAuthHandler.java (hunk 2 above)
Comment (ayshsandu) [reply component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/PrivateKeyJWTClientAuthHandler.java]: Added an inline comment

File: component/client-handler/pom.xml [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+            </plugins>
+        </pluginManagement>
+        <plugins>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-compiler-plugin</artifactId>
+                <version>2.0</version>
+                <configuration>
+                    <source>1.7</source>
+                    <target>1.7</target>
```
Comment (sagara-gunathunga) [inline component/client-handler/pom.xml]: Can you check  whether  we move to Java 8 or not ? if we moved we have to change this.

File: component/client-handler/pom.xml [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+                    <version>2.3.5</version>
+                    <extensions>true</extensions>
+                </plugin>
+            </plugins>
+        </pluginManagement>
+        <plugins>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-compiler-plugin</artifactId>
+                <version>2.0</version>
```
Comment (sagara-gunathunga) [inline component/client-handler/pom.xml]: Versioning  etc. should go to <pluginManagement> section.

File: component/client-handler/pom.xml [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+                            org.wso2.carbon.identity.oauth2.token.handler.clientauth.jwt.*;
+                        </Export-Package>
+                        <DynamicImport-Package>*</DynamicImport-Package>
+                    </instructions>
+                </configuration>
+            </plugin>
+            <plugin>
+                <groupId>org.apache.felix</groupId>
+                <artifactId>maven-scr-plugin</artifactId>
+                <version>1.7.2</version>
```
Comment (sagara-gunathunga) [inline component/client-handler/pom.xml]: Versioning etc. should go to section.

File: component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/cache/JWTCache.java [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+//        super.addToCache(key, entry);
+//    }
+//
+//    public CacheEntry getValueFromCache(String key) {
+//        return (CacheEntry) super.getValueFromCache(key);
+//    }
+//
+//    public void clearCacheEntry(String key) {
+//        super.clearCacheEntry(key);
+//    }
```
Comment (sagara-gunathunga) [inline component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/cache/JWTCache.java]: Remove commented lines.

File: component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/validator/JWTValidator.java [hunk 8]
```diff
@@ -0,0 +1,657 @@
...
+                        boolean enableJTICache, String validAudience, String validIssuer) {
+        this.notAcceptBeforeTimeInMins = rejectBeforePeriod;
+        this.preventTokenReuse = preventTokenReuse;
+        this.enableJTICache = enableJTICache;
+        this.validAudience = validAudience;
+        this.validIssuer = validIssuer;
+        this.jwtStorageManager = new JWTStorageManager();
+        if (enableJTICache) {
+            this.jwtCache = JWTCache.getInstance();
+        } else {
```
Comment (sagara-gunathunga) [inline component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/validator/JWTValidator.java]: I think we agreed we don't need these blocks

File: component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/validator/JWTValidator.java [hunk 9]
```diff
@@ -0,0 +1,657 @@
...
+            String errorMsg = "Error getting the tenant ID for the tenant domain : " + tenantDomain;
+            throw new IdentityOAuth2Exception(errorMsg, e);
+        }
+
+        KeyStoreManager keyStoreManager;
+        // get an instance of the corresponding Key Store Manager instance
+        keyStoreManager = KeyStoreManager.getInstance(tenantId);
+        KeyStore keyStore;
+        try {
+            if (tenantId != -1234) {// for tenants, load key from their generated key store
```
Comment (sagara-gunathunga) [inline component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/validator/JWTValidator.java]: Need a constant here

File: component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/validator/JWTValidator.java [hunk 10]
```diff
@@ -0,0 +1,657 @@
...
+        }
+
+        String alg = signedJWT.getHeader().getAlgorithm().getName();
+        if (isEmpty(alg)) {
+            return false;
+        } else {
+            if (log.isDebugEnabled()) {
+                log.debug("Signature Algorithm found in the JWT Header: " + alg);
+            }
+            if (alg.indexOf("RS") == 0) {
```
Comment (sagara-gunathunga) [inline component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/validator/JWTValidator.java]: Need a constant here

File: component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/validator/JWTValidator.java [hunk 11]
```diff
@@ -0,0 +1,657 @@
...
+     * @param expirationTime      Expiration time
+     * @param currentTimeInMillis Current time
+     * @param timeStampSkewMillis Time skew
+     * @return true or false
+     */
+    public boolean validateExpirationTime(Date expirationTime, long currentTimeInMillis,
+                                          long timeStampSkewMillis) throws IdentityOAuth2Exception {
+        long expirationTimeInMillis = expirationTime.getTime();
+        if ((currentTimeInMillis + timeStampSkewMillis) > expirationTimeInMillis) {
+            return logAndReturnFalse("JSON Web Token is expired." +
```
Comment (sagara-gunathunga) [inline component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/validator/JWTValidator.java]: We agreed to use StringBuilder here

File: component/client-handler/src/test/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/PrivateKeyJWTClientAuthHandlerTest.java [hunk 12]
```diff
@@ -0,0 +1,365 @@
...
+import org.wso2.carbon.identity.oauth2.token.handler.clientauth.jwt.util.JWTTestUtil;
+import org.wso2.carbon.identity.oauth2.token.handler.clientauth.jwt.validator.JWTValidator;
+import org.wso2.carbon.identity.testutil.ReadCertStoreSampleUtil;
+import org.wso2.carbon.idp.mgt.internal.IdpMgtServiceComponentHolder;
+import org.wso2.carbon.user.api.UserRealm;
+import org.wso2.carbon.user.core.service.RealmService;
+
+import java.io.ByteArrayInputStream;
+import java.lang.reflect.Field;
+import java.security.*;
```
Comment (sagara-gunathunga) [inline component/client-handler/src/test/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/PrivateKeyJWTClientAuthHandlerTest.java]: Why we still use * imports here

File: component/client-handler/src/test/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/PrivateKeyJWTClientAuthHandlerTest.java [hunk 13]
```diff
@@ -0,0 +1,365 @@
...
+        injectToSingletons = {JWTServiceComponent.class})
+@WithKeyStore
+public class PrivateKeyJWTClientAuthHandlerTest {
+
+    public static final String TEST_CLIENT_ID_1 = "KrVLov4Bl3natUksF2HmWsdw684a";
+    public static final String TEST_SECRET_1 = "testSecret1";
+    public static final String OAUTH_CLIENT_ASSERTION_TYPE_JWT_BEARER = "urn:ietf:params:oauth:client-assertion-type:jwt-bearer";
+
+    private KeyStore clientKeyStore;
+    private String CERTIFICATE = "MIIDVzCCAj+gAwIBAgIEN+6m4zANBgkqhkiG9w0BAQsFADBcMQswCQYDVQQGEwJGUjEMMAoGA1UE\n" +
```
Comment (sagara-gunathunga) [inline component/client-handler/src/test/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/PrivateKeyJWTClientAuthHandlerTest.java]: Hard coding certificate  is the right  approach to use ?

File: component/client-handler/src/test/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/PrivateKeyJWTClientAuthHandlerTest.java [hunk 14]
```diff
@@ -0,0 +1,365 @@
...
+    }
+
+    @Test(dataProvider = "provideOAuthTokenReqMessageContext")
+    public void testAuthenticateClient(Object oAuthTokenReqMessageContext, boolean canAuthenticate,
+                                       Object properties, boolean expectedValue, String errorMsg)
+            throws InvalidOAuthClientException, IdentityOAuth2Exception, IdentityOAuthAdminException {
+
+        PrivateKeyJWTClientAuthHandler privateKeyJWTClientAuthHandler = new PrivateKeyJWTClientAuthHandler();
+        privateKeyJWTClientAuthHandler.init((Properties) properties);
+//        testClass.init((Properties) properties);
```
Comment (sagara-gunathunga) [inline component/client-handler/src/test/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/PrivateKeyJWTClientAuthHandlerTest.java]: ??

File: component/client-handler/src/test/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/PrivateKeyJWTClientAuthHandlerTest.java (hunk 13 above)
Comment (ayshsandu) [reply component/client-handler/src/test/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/PrivateKeyJWTClientAuthHandlerTest.java]: These are in unit tests. want to validate test case with some certificate.

File: component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/cache/JWTCache.java (hunk 7 above)
Comment (ayshsandu) [reply component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/cache/JWTCache.java]: Done

File: component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/validator/JWTValidator.java (hunk 9 above)
Comment (ayshsandu) [reply component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/validator/JWTValidator.java]: done

File: component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/validator/JWTValidator.java (hunk 10 above)
Comment (ayshsandu) [reply component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/validator/JWTValidator.java]: done

File: component/client-handler/src/test/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/PrivateKeyJWTClientAuthHandlerTest.java (hunk 14 above)
Comment (ayshsandu) [reply component/client-handler/src/test/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/PrivateKeyJWTClientAuthHandlerTest.java]: removed commented lines

File: component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/validator/JWTValidator.java (hunk 11 above)
Comment (ayshsandu) [reply component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/validator/JWTValidator.java]: refactored

**Pull request merged**